SG-12960 doc edits and additions for direct upload to s3

zoeglynn opened the pull request:
Biggest changes are:

<img width="488" alt="Screen_Shot_2019-06-26_at_8_51_21_AM" src="https://user-images.githubusercontent.com/227512/60195315-b7509400-97ef-11e9-846a-fa42159cf072.png">

<img width="567" alt="Screen_Shot_2019-06-26_at_8_49_03_AM" src="https://user-images.githubusercontent.com/227512/60195722-7442f080-97f0-11e9-8336-626844294f5d.png">

jfboismenu commented on shotgun_api3/shotgun.py:
```diff
@@ -2085,6 +2088,42 @@ def share_thumbnail(self, entities, thumbnail_path=None, source_entity=None,
+
+            def share(jpeg_file):
+               # Set your project
+               P = {"type":"Project","id":project_id}
```
Let's avoid one letter variables. `project` would be cleaner.

jfboismenu commented on shotgun_api3/shotgun.py:
```diff
@@ -2085,6 +2088,42 @@ def share_thumbnail(self, entities, thumbnail_path=None, source_entity=None,
+               P = {"type":"Project","id":project_id}
+
+               # Create a version
+               props = {"code":"Test Version","project":P}
```
Seems like the sample code is inconsistent with spacing around `,` and `:`. Might I recommend always having a space after those, as per pep8 style guide?

jfboismenu commented on shotgun_api3/shotgun.py:
```diff
@@ -2085,6 +2088,42 @@ def share_thumbnail(self, entities, thumbnail_path=None, source_entity=None,
+               props = {"code":"Test Asset","project":P}
+               asset_id = sg.create("Asset", props).get("id")
+
+               # Now upload attach the thumbnail to the version
```
~attach~

jfboismenu commented on shotgun_api3/shotgun.py:
```diff
@@ -2085,6 +2088,42 @@ def share_thumbnail(self, entities, thumbnail_path=None, source_entity=None,
+               v_id = sg.create("Version", props).get("id")
+
+               # Create an asset
+               props = {"code":"Test Asset","project":P}
```
Missing spaces after : and ,

jfboismenu commented on shotgun_api3/shotgun.py:
```diff
@@ -2085,6 +2088,42 @@ def share_thumbnail(self, entities, thumbnail_path=None, source_entity=None,
+                       time.sleep(5)
+
+               # Validate the thumbnail was indeed shared
+               e = sg.find_one('Asset',[['id', 'is',  asset_id]], ['id', 'image'])
```
Missing spaces after ,

zoeglynn commented on docs/cookbook/examples/basic_upload_thumbnail_version.rst:
```diff
@@ -24,4 +24,21 @@ will be thrown and you'll see the error details.
+ __ https://developer.shotgunsoftware.com/python-api/reference.html#shotgun_api3.shotgun.Shotgun.share_thumbnail
+ __ https://github.com/shotgunsoftware/shotgunEvents/wiki/Technical_Overview#event-log-entries-for-thumbnails
+
+.. note:: Shotgun may require transient thumbnails in future features. For this reason, we highly
```
Yes let me clarify the intent of this note.

zoeglynn replied:
I'm happy with 'transient' here in that it implies a temporary value that is expected to change to something else

zoeglynn commented:
Have attached the docs from sphinx build to make review easier - 
[python-api.zip](https://github.com/shotgunsoftware/python-api/files/3777395/python-api.zip)
Notable changes:
* Added "Interpreting Image Field Strings' section under Data Types, to match REST docs (https://developer.shotgunsoftware.com/rest-api/#interpreting-image-field-strings)
* For `Shotgun.share_thumbnail`:
  - removed the 'Pending thumbnail retry example' section, since it could encourage poor practice. 
  - document new `ShotgunThumbnailNotReady` error class.

zoeglynn commented on docs/cookbook/examples/basic_upload_thumbnail_version.rst:
```diff
@@ -24,4 +24,21 @@ will be thrown and you'll see the error details.
+
+.. note:: Shotgun may require transient thumbnails in future features. For this reason, we highly
+	  recommend using the prefix part of the placeholder URL (``/images/status/transient``) to
+	  detect any transient thumbnails, rather than using the full URL.
```
Content no longer added to this file; but also ok with the use of "transient" to describe the path returned.

jfboismenu commented on docs/cookbook/attachments.rst:
```diff
@@ -76,12 +76,17 @@ descriptions of each field.
 
+- **processing_status** (:obj:`str`):
+    Reflects the status of the attachment (File entity).
+    When processing the thumbnail, this field is set to ‘Thumbnail Pending’.
```
Should we explain how to use this field? I thought we were going to have a helper method that you could pass an image to and it would tell you if the image was ready or not?

jfboismenu replied:
Sorry, Github says I'm requesting for changes. I don't know why that is. I'm pretty sure I clicked "Comment" instead of Request Change.
